{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Imports and loading config data"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "from itertools import combinations_with_replacement\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import dcor\n",
    "import yaml\n",
    "import os\n",
    "from pathlib import Path\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from statsmodels.graphics.tsaplots import plot_acf, plot_pacf, acf\n",
    "from matplotlib.collections import PolyCollection, LineCollection\n",
    "from matplotlib.lines import Line2D\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "# Load configuration file\n",
    "with open('params_icgcl.yaml', 'rb') as f:\n",
    "    conf = yaml.safe_load(f.read())\n",
    "# Set seaborn style\n",
    "sns.set_style(style='white')\n",
    "# Plot settings\n",
    "plt.rc('text', usetex=True)\n",
    "plt.rc('font', **{'family': \"sans-serif\"})\n",
    "plt.rc('text.latex', preamble=r\"\\usepackage{amsmath}\" r\"\\usepackage{amstext}\")\n",
    "plt.rcParams[\"axes.linewidth\"] = 2.50\n",
    "plt.rcParams['xtick.major.size'] = 20\n",
    "plt.rcParams['ytick.major.size'] = 20\n",
    "fsz = 28"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "# Load general settings from the config file\n",
    "save_img = conf['settings']['save_img']\n",
    "active_dataset = conf[\"settings\"][\"active_dataset\"]\n",
    "dataset_config = conf[\"datasets\"][active_dataset]\n",
    "\n",
    "# Load dataset-specific parameters\n",
    "ec_hc_mask_folder = dataset_config[\"ec_hc_mask_folder\"]\n",
    "input_file = dataset_config[\"input_file\"]\n",
    "input_folder = dataset_config[\"input_folder\"]\n",
    "output_folder = dataset_config[\"output_folder\"]\n",
    "rep_list = dataset_config[\"rep_list\"]"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "df_list_treat = {'WRN ASO': 'c', 'HGPS ASO': '#66CDAA', 'WT': 'g', 'WRN SCR': 'r', 'WRN NT':'#D70040', 'HGPS SCR': 'm', 'HGPS NT':'#8B008B'}"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data Preprocessing and Analysis functions"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "def data_preprocessing(df, expt_list):\n",
    "    normalized_df = df[['Gene_ID', 'StartPos', 'EndPos']].copy()\n",
    "    normalized_df[expt_list] = df[expt_list].div(df[expt_list].sum(axis=0), axis=1)\n",
    "    return normalized_df.reset_index(drop=True)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Correlation Analysis Functions"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "def get_pos_val(df,expt):\n",
    "    tp_gene = np.array(df[expt]) # Here are the transcription values\n",
    "    pos_start_end = df[['StartPos','EndPos']].values.tolist()\n",
    "    return  pos_start_end, tp_gene"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "def convert_to_sorted_dict(pos_tuple,tp_gene):\n",
    "    pos_value = dict()\n",
    "    for i in range(len(pos_tuple)):\n",
    "        start_end = pos_tuple[i]\n",
    "        start_end=(start_end[0],start_end[1])\n",
    "        value = tp_gene[i]\n",
    "        pos_value[start_end] = value\n",
    "    return dict(sorted(pos_value.items()))"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "def get_distance(pos_tuple,tp_gene):\n",
    "    bin = dict()\n",
    "    # Here the pos_value is a dictionary with key as the position and value as the transcription.\n",
    "    # The pos value is sorted based on the position\n",
    "    pos_value = convert_to_sorted_dict(pos_tuple,tp_gene)\n",
    "    prev_key_1 = \"\"\n",
    "    # we generate now combinations like\n",
    "    # a b\n",
    "    # a c\n",
    "    # a d\n",
    "    # b c\n",
    "    # b d\n",
    "    #....\n",
    "    # where a and b would be the start positions (or keys in the dictionary)\n",
    "    # now we calculate the distance based on the relative index and not on the start position\n",
    "    # Hence we keep i and j and counters for the indices\n",
    "    i=-1\n",
    "    j=-1\n",
    "    for key_1, key_2 in combinations_with_replacement(pos_value.keys(), r = 2):\n",
    "        if prev_key_1!=key_1:\n",
    "            i=i+1\n",
    "            prev_key_1=key_1\n",
    "            j=i\n",
    "        else:\n",
    "            j=j+1\n",
    "        value_1 = pos_value[key_1]\n",
    "        value_2 = pos_value[key_2]\n",
    "        tuple_tp = (value_1,value_2)\n",
    "        bin.setdefault(j-i, []).append(tuple_tp)\n",
    "    return bin"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "def calculate_binwise_correlation(bin):\n",
    "    binwise_correlation = {}\n",
    "    for bin_idx,tuple_list in bin.items():\n",
    "        x_y = list(map(list, zip(*tuple_list)))\n",
    "        x = x_y[0]\n",
    "        y = x_y[1]\n",
    "        cor = dcor.distance_correlation(np.array(x),np.array(y))\n",
    "        binwise_correlation[bin_idx] = cor\n",
    "    return binwise_correlation"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "def get_acf_pacf_intercept(dict_values):\n",
    "    a, ci = acf(x=np.array(list(dict_values.values())), nlags=len(dict_values)-1, alpha=0.05)\n",
    "    # Key step\n",
    "    centered_ci = ci - a[:,None]\n",
    "    outside = np.abs(a) >= centered_ci[:,1]\n",
    "    intercept = (np.where(outside == False)[0][0])-1\n",
    "    return intercept"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "def get_and_plot_acf_pacf(dict_values_ec,dict_values_hc, save_img,save_plot_path):\n",
    "    # intercept = get_acf_pacf_intercept(dict_values)\n",
    "    intercept_corr_ec, intercept_corr_hc =  save_acf_pacf_plot(dict_values_ec, dict_values_hc, save_plot_path, save_img)  \n",
    "    return intercept_corr_ec, intercept_corr_hc"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Visualization Functions"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "def save_cluster(df, save_plot_path):\n",
    "    fig, ax = plt.subplots()\n",
    "    fig.set_size_inches(12, 10)\n",
    "    fig.set_dpi(150)\n",
    "    df['index'] = range(1, len(df)+1)\n",
    "    df['log_index'] = np.log10(df['index'])\n",
    "    palette = [ \"#008000\",\"#FF0000\"]\n",
    "    for i in df.columns[0:3]:\n",
    "        sns.scatterplot(ax=ax, x=\"index\", y=i, data=df, hue=\"labels\",s=fsz+50,palette=sns.color_palette(palette, 3),linewidth=0, alpha = 0.7)\n",
    "\n",
    "    ax.set_ylabel(r'$\\alpha_t$', fontsize = fsz, color='k')\n",
    "    ax.set_xlabel('gene expression rank ($\\mathrm{r}_t$)', fontsize=fsz, color='k')\n",
    "    ax.tick_params('both',which='major', length=7,labelsize=fsz)\n",
    "    ax.tick_params('both',which='minor', length=7,labelsize=fsz)\n",
    "    ax.set(xscale= 'log')\n",
    "    ax.set(yscale= 'log')\n",
    "    labels = ['High abundance transcript (HAT)','Low abundance transcript (LAT)']\n",
    "    legend_elements = [Line2D([0], [0], marker='o',color = 'r', lw=0, markerfacecolor='r', label=labels[0],\n",
    "                                  markersize=8),\n",
    "                        Line2D([0], [0], marker='o',color = 'g', lw=0, markerfacecolor='g', label=labels[1],\n",
    "                                  markersize=8)]\n",
    "\n",
    "    ax.legend(handles=legend_elements,markerscale=2, loc='upper right', borderaxespad=0.05, fontsize = fsz-4)\n",
    "    plt.savefig(save_plot_path)\n",
    "    plt.close()"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "def plot_acf_colors(ax, markercolor=\"red\", linecolor=\"black\", facecolor=\"red\", barcolor=\"red\", linewidth=1):\n",
    "    \"\"\"utility function to get some control over colors with  plot_acf()\"\"\"\n",
    "    for item in ax.collections:\n",
    "        # change the color of the confidence interval\n",
    "        if type(item) == PolyCollection:\n",
    "            item.set_facecolor(facecolor)\n",
    "        # change the color of the vertical lines\n",
    "        if type(item) == LineCollection:\n",
    "            item.set_color(barcolor)\n",
    "    # change the color of the markers\n",
    "    [line.get_label() for line in ax.lines]\n",
    "    for item in ax.lines:\n",
    "        item.set_color(markercolor)\n",
    "    # change the color of the horizontal lines\n",
    "    ax.lines[0].set_color(linecolor)\n",
    "    ax.lines[0].set_linewidth(linewidth)\n",
    "    return ax"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "# This is used to plot Fig 1\n",
    "def save_acf_pacf_plot(dict_values_ec,dict_values_hc,save_plot_path,save_img):\n",
    "    intercept_ec = get_acf_pacf_intercept(dict_values_ec)\n",
    "    intercept_hc = get_acf_pacf_intercept(dict_values_hc)\n",
    "    if save_img:\n",
    "        fig, ax = plt.subplots(1,2)\n",
    "        fig.set_size_inches(14, 7)\n",
    "        fig.set_dpi(150)\n",
    "\n",
    "        plot_acf(x=np.array(list(dict_values_ec.values())),lags=len(dict_values_ec)-1,ax=ax[0])\n",
    "        plot_acf(x=np.array(list(dict_values_hc.values())),lags=len(dict_values_hc)-1,ax=ax[1])\n",
    "\n",
    "        ax[0].tick_params('both',which='major', length=7,labelsize=fsz)\n",
    "        ax[0].tick_params('both',which='minor', length=7,labelsize=fsz)\n",
    "        ax[0].set_ylabel('$\\mathcal{R}(\\ell)$', fontsize=fsz, color='k')\n",
    "        ax[0].set_xlabel('$\\ell$', fontsize=fsz, color='k')\n",
    "\n",
    "        ax[1].tick_params('both',which='major', length=7,labelsize=fsz)\n",
    "        ax[1].tick_params('both',which='minor', length=7,labelsize=fsz)\n",
    "        ax[1].set_xlabel('$\\ell$', fontsize=fsz, color='k')\n",
    "        ax[0].set_ylim(bottom=-0.5, top=1)\n",
    "\n",
    "        ax[0] = plot_acf_colors(ax[0],markercolor=\"red\", linecolor=\"black\", facecolor=\"red\", barcolor=\"red\", linewidth=1)\n",
    "        ax[1] = plot_acf_colors(ax[1],markercolor=\"green\", linecolor=\"black\", facecolor=\"green\", barcolor=\"green\", linewidth=1)\n",
    "\n",
    "        for i in range(1,len(ax)):\n",
    "            ax[i].set_ylim( ax[0].get_ylim() ) # align axes\n",
    "            ax[i].set_yticks([]) # set ticks to be empty (no ticks, no tick-labels)\n",
    "\n",
    "        ax[0].vlines(x=intercept_ec, ymin=-0.5, ymax=1, colors=\"black\", ls='--', lw=2, alpha=1)\n",
    "        ax[1].vlines(x=intercept_hc, ymin=-0.5, ymax=1, colors=\"black\", ls='--', lw=2, alpha=1)\n",
    "        plt.savefig(save_plot_path, format='pdf', bbox_inches='tight')\n",
    "        plt.close(fig)\n",
    "    return intercept_ec, intercept_hc"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "# This is used to plot Fig 2\n",
    "def get_and_plot_dcor(bin_ec_idx_cor_dict,bin_hc_idx_cor_dict,cm_ec,cm_hc,save_img,save_plot_path):\n",
    "    if save_img:\n",
    "        fig, ax = plt.subplots(1,1)\n",
    "        fig.set_size_inches(12, 8)\n",
    "        fig.set_dpi(520)\n",
    "        plt.plot(list(bin_ec_idx_cor_dict.keys()),list(bin_ec_idx_cor_dict.values()),color = cm_ec,linewidth = 3)\n",
    "        plt.plot(list(bin_hc_idx_cor_dict.keys()),list(bin_hc_idx_cor_dict.values()),color = cm_hc, linewidth = 3)\n",
    "        ax.tick_params('both',which='major', length=7,labelsize=fsz)\n",
    "        ax.tick_params('both',which='minor', length=7,labelsize=fsz)\n",
    "        ax.set_ylabel('$\\mathcal{C}(\\Delta \\ell)$', fontsize=fsz, color='k')\n",
    "        ax.set_xlabel('$\\Delta \\ell$', fontsize=fsz, color='k')\n",
    "\n",
    "        ax.tick_params('both',which='major', length=7,labelsize=fsz)\n",
    "        # ax.set_xlim(left=0, right=3500)\n",
    "        plt.savefig(save_plot_path, format='pdf', bbox_inches='tight')\n",
    "        plt.close(fig)\n",
    "    return "
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "# This is used to plot Fig 7\n",
    "def plot_for_rank(df,save_plot_path):\n",
    "    fig, ax = plt.subplots()\n",
    "    fig.set_size_inches(12, 10)\n",
    "    fig.set_dpi(150)\n",
    "    df_list_treat = {'ASO': 'c', 'WT': 'g', 'WRN': 'r', 'HGPS': 'm'} # Creating a dictionary for the color for different treatments\n",
    "    ind =[]\n",
    "    ind_1 =[]\n",
    "    gene_val = []\n",
    "    gene_val_1 = []\n",
    "    expt_val = []\n",
    "    gene_ref = 'ENSG00000117595'\n",
    "    gene_ref_1 = 'ENSG00000162433'\n",
    "    for col in df.columns[1:]:\n",
    "        df_new = df[['Gene_ID', col]].sort_values(by = col, ascending=False,ignore_index = True).reset_index(drop=True)\n",
    "        x = np.linspace(1,len(df_new), len(df_new))\n",
    "\n",
    "        ind.append(df_new.index[df_new['Gene_ID'] == gene_ref].tolist())\n",
    "        ind_1.append(df_new.index[df_new['Gene_ID'] == gene_ref_1].tolist())\n",
    "        gene_val.append(df_new[df_new['Gene_ID'] == gene_ref][col].tolist())\n",
    "        gene_val_1.append(df_new[df_new['Gene_ID'] == gene_ref_1][col].tolist())\n",
    "        expt_val.append(col)\n",
    "        temp = [i for i, s in enumerate(df_list_treat) if s in col]\n",
    "        ax.plot(x,df_new[col].to_numpy(),'*', label = col, color = list(df_list_treat.items())[int(temp[0])][1],alpha=1.0)\n",
    "        ax.set_ylabel(r'$\\alpha_t$', fontsize = fsz+6, color='k')\n",
    "        ax.set_xlabel('gene expression rank ($\\mathrm{r}_t$)', fontsize=fsz+6, color='k')\n",
    "        ax.tick_params('both',which='major', length=7,labelsize=fsz)\n",
    "        ax.tick_params('both',which='minor', length=7,labelsize=fsz)\n",
    "        # ax.set_ylim(bottom=1e-5, top=1e-2)\n",
    "        ax.set(xscale= 'log')\n",
    "        ax.set(yscale= 'log')\n",
    "\n",
    "    ind = [item for sublist in ind for item in sublist]\n",
    "    ind_1 = [item for sublist in ind_1 for item in sublist]\n",
    "    gene_val = [item for sublist in gene_val for item in sublist]\n",
    "    gene_val_1 = [item for sublist in gene_val_1 for item in sublist]\n",
    "\n",
    "    df_sort = pd.DataFrame(list(zip(ind, gene_val, expt_val)), columns=['ind','val','expt'])\n",
    "    df_sort_1 = pd.DataFrame(list(zip(ind_1, gene_val_1, expt_val)), columns=['ind_1','val_1','expt'])\n",
    "    \n",
    "    col_order = ['WT_1_1','WT_2_1','WT_3_1', 'HGPS_917_L1_ASO_1','HGPS_478_L1_ASO_1','HGPS_297_L1_ASO_1', 'WRN_1_L1_ASO_1','WRN_2_L1_ASO_1','WRN_3_L1_ASO_1','HGPS_917_SCR_1','HGPS_478_SCR_1','HGPS_297_SCR_1','WRN_1_SCR_1','WRN_2_SCR_1','WRN_3_SCR_1','HGPS_917_NT_1','HGPS_478_NT_1','HGPS_297_NT_1','WRN_1_NT_1','WRN_2_NT_1','WRN_3_NT_1']\n",
    "    \n",
    "    # Create a categorical column with the custom order\n",
    "    df_sort['expt_cat'] = pd.Categorical(df_sort['expt'], categories=col_order, ordered=True)\n",
    "    \n",
    "    # Sort the dataframe based on the new categorical column\n",
    "    df_sorted = df_sort.sort_values('expt_cat')\n",
    "    # If you want to drop the temporary categorical column\n",
    "    df_sorted = df_sorted.drop('expt_cat', axis=1)    \n",
    "    # Reset the index if needed\n",
    "    df_sort = df_sorted.reset_index(drop=True)\n",
    "\n",
    "    df_sort = pd.merge(df_sort, df_sort_1, on=\"expt\")\n",
    "    handles, labels = ax.get_legend_handles_labels()\n",
    "    ## specify order of items in legend\n",
    "    order = [0,1,2,3,4,5,6,7,8,9,10,11,18,19,20,12,13,14,15,16,17]\n",
    "        ## add legend to plot\n",
    "    ax.legend([handles[idx] for idx in order],[labels[idx] for idx in order],markerscale=2, bbox_to_anchor=(1.01, 1), loc=2, borderaxespad=0.05, fontsize = fsz-10)\n",
    "    x = np.linspace(1,len(df_sort), len(df_sort))\n",
    "    y = df_sort['ind']\n",
    "    y1 = df_sort['ind_1']\n",
    "    left, bottom, width, height = [0.205, 0.33, 0.33, 0.33]\n",
    "    ax1 = fig.add_axes([left, bottom, width, height])\n",
    "\n",
    "    ax1.plot(x,y,'ob')\n",
    "    ax1.set_xticks(x)\n",
    "    ax1.set_xticklabels(list(df_sort['expt']))\n",
    "    ax1.grid(False)\n",
    "    ax1.set_ylabel('$\\mathrm{r}_t$', fontsize=fsz, color='k')\n",
    "    ax1.tick_params('both',which='major', length=7,labelsize=fsz-12)\n",
    "    ax1.tick_params('both',which='minor', length=7,labelsize=fsz-12)\n",
    "    plt.xticks(rotation=90)\n",
    "\n",
    "    ax2 = ax1.twinx()\n",
    "    ax2.plot(y1,'or')\n",
    "    # Setting the color for the axis\n",
    "    color = []\n",
    "    for col in df_sort['expt']:\n",
    "        temp = [i for i, s in enumerate(df_list_treat) if s in col]\n",
    "        color.append(list(df_list_treat.items())[temp[0]][1])\n",
    "    i = 0\n",
    "    for xtick in ax1.get_xticklabels():\n",
    "        xtick.set_color(color[i])\n",
    "        i = i + 1\n",
    "    ax2.tick_params('y',which='major', length=7,labelsize=fsz-12)\n",
    "    ax2.tick_params('y',which='minor', length=7,labelsize=fsz-12)\n",
    "    ax1.yaxis.label.set_color('b')\n",
    "    ax2.spines[\"left\"].set_edgecolor('b')\n",
    "    ax1.tick_params(axis='y', colors='b')\n",
    "    ax2.yaxis.label.set_color('r')\n",
    "    ax2.spines[\"right\"].set_edgecolor('r')\n",
    "    ax2.tick_params(axis='y', colors='r')\n",
    "    plt.savefig(save_plot_path, bbox_inches='tight')\n",
    "    plt.close()"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Function to generate output Summary"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "def generate_summary_file(output_folder):\n",
    "    auto_cor_ratio_df = pd.DataFrame()\n",
    "    total_genes_df = pd.DataFrame()\n",
    "    counter=1\n",
    "    for root, dirs, files in os.walk(output_folder):\n",
    "        for dir in dirs:\n",
    "            file = os.path.join(output_folder,dir,dir+\"_auto_cor_ratio.csv\")\n",
    "            file_1 = os.path.join(output_folder,dir,dir+\"_ec_genes_df.csv\")\n",
    "            df = pd.read_csv(file)\n",
    "            df_1 = pd.read_csv(file_1)\n",
    "            if counter == 1:\n",
    "                expt_list = list(df.columns)\n",
    "                auto_cor_ratio_df = pd.DataFrame(columns=expt_list)\n",
    "                total_genes_df = pd.DataFrame(columns=expt_list)\n",
    "                counter = counter+1\n",
    "            auto_cor_ratio_df = pd.concat([auto_cor_ratio_df, df],ignore_index=True)\n",
    "            total_genes_df = pd.concat([total_genes_df, df_1],ignore_index=True)\n",
    "    auto_cor_ratio_df.to_csv(os.path.join(output_folder,\"auto_cor_ratio.csv\"),index=False)\n",
    "    total_genes_df.to_csv(os.path.join(output_folder,\"total_genes_df.csv\"),index=False)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Main Processing Pipeline"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "df = pd.read_csv(input_file)\n",
    "df['Chromosome']=df['Chromosome'].astype(str)\n",
    "df.sort_values(\"Gene_ID\",inplace=True)\n",
    "count = 0\n",
    "expt_list = [item for sublist in rep_list for item in sublist]\n",
    "\n",
    "for chromosome,data in df.groupby('Chromosome'):\n",
    "    if chromosome.isalnum() and chromosome != 'MT' and chromosome != 'Y':\n",
    "        chromosome_output_folder = os.path.join(output_folder,\"chromosome_\"+str(chromosome))\n",
    "        Path(chromosome_output_folder).mkdir(parents=True, exist_ok=True)\n",
    "        auto_cor_ratio_df = pd.DataFrame(columns= ['chromosome_id'] + expt_list)\n",
    "        total_genes_df = pd.DataFrame(columns= ['chromosome_id'] + expt_list)\n",
    "        normalized_df = data_preprocessing(data,expt_list)\n",
    "        auto_cor_ratio_list = [chromosome]\n",
    "        total_genes_list = [chromosome]    \n",
    "        precomp_file = os.path.join(ec_hc_mask_folder, f\"chromosome_{chromosome}\" ,f\"chromosome_{chromosome}_ec_hc_mask_xy.csv\")\n",
    "        if not os.path.exists(precomp_file):\n",
    "            print(f\"Precomputed HAT/LAT data not found for chromosome {chromosome}. Skipping...\")\n",
    "            continue\n",
    "\n",
    "        new_df = pd.read_csv(precomp_file)\n",
    "        data_hi_folder = os.path.join(ec_hc_mask_folder,f\"chromosome_{chromosome}\" ,\"data_hi\")\n",
    "        data_lo_folder = os.path.join(ec_hc_mask_folder,f\"chromosome_{chromosome}\" , \"data_lo\")\n",
    "        for rep in rep_list:\n",
    "            data_hi_file = os.path.join(data_hi_folder, f\"chromosome_{chromosome}_rep_{rep[0]}_data_hi.csv\")\n",
    "            data_lo_file = os.path.join(data_lo_folder, f\"chromosome_{chromosome}_rep_{rep[0]}_data_lo.csv\")\n",
    "            \n",
    "            if os.path.exists(data_hi_file) and os.path.exists(data_lo_file):\n",
    "                data_hi = pd.read_csv(data_hi_file, index_col=0)\n",
    "                data_lo = pd.read_csv(data_lo_file, index_col=0)\n",
    "            else:\n",
    "                print(f\"Data files not found for replicate {rep[0]} in chromosome {chromosome}. Exiting...\")\n",
    "                break\n",
    "    \n",
    "            for expt in rep:\n",
    "                print(\"Processing Chromosome :\"+str(chromosome)+\" and experiment \"+str(expt))\n",
    "                positions_ec,tp_gene_ec = get_pos_val(data_hi,expt)\n",
    "                positions_hc,tp_gene_hc = get_pos_val(data_lo,expt)\n",
    "                bin_ec_idx_dis_dict = get_distance(positions_ec,tp_gene_ec)\n",
    "                bin_hc_idx_dis_dict = get_distance(positions_hc,tp_gene_hc)\n",
    "                bin_ec_idx_cor_dict = calculate_binwise_correlation(dict(sorted(bin_ec_idx_dis_dict.items())))    \n",
    "                bin_hc_idx_cor_dict = calculate_binwise_correlation(dict(sorted(bin_hc_idx_dis_dict.items())))\n",
    "                # Let's calculate the intercept for both ec and hc\n",
    "                intercept_corr_ec, intercept_corr_hc = get_and_plot_acf_pacf(bin_ec_idx_cor_dict, bin_hc_idx_cor_dict, save_img, os.path.join(chromosome_output_folder,(expt+\"_autocorr_HAT_LAT.pdf\")))\n",
    "                get_and_plot_dcor(bin_ec_idx_cor_dict,bin_hc_idx_cor_dict,\"red\",\"green\",save_img,os.path.join(chromosome_output_folder,(expt+\"_dcor_HAT_LAT.pdf\"))) \n",
    "                \n",
    "                scaling_factor = (len(positions_ec)+len(positions_hc))\n",
    "                intercept_corr_ec = (intercept_corr_ec+1) * 1000 / scaling_factor\n",
    "                auto_cor_ratio_list.append(intercept_corr_ec)\n",
    "                total_genes_list.append((len(positions_hc)+len(positions_ec)))\n",
    "                print(\"intercept_corr_ec\", intercept_corr_ec,scaling_factor)\n",
    "           \n",
    "        auto_cor_ratio_df = auto_cor_ratio_df.append(pd.DataFrame([auto_cor_ratio_list], columns=['chromosome_id'] + expt_list), ignore_index=True)\n",
    "        auto_cor_ratio_df.to_csv(os.path.join(chromosome_output_folder,\"chromosome_\"+str(chromosome)+\"_auto_cor_ratio.csv\"),index=False)\n",
    "        total_genes_df = total_genes_df.append(pd.DataFrame([total_genes_list], columns=['chromosome_id'] + expt_list), ignore_index=True)\n",
    "        total_genes_df.to_csv(os.path.join(chromosome_output_folder,\"chromosome_\"+str(chromosome)+\"_ec_genes_df.csv\"),index=False)\n",
    "generate_summary_file(output_folder)"
   ],
   "outputs": [],
   "execution_count": null
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
