{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Imports and loading config data"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "import pandas as pd\n",
    "import math\n",
    "from sklearn import mixture\n",
    "import pymannkendall as mk\n",
    "import numpy as np\n",
    "import yaml\n",
    "import os\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from pathlib import Path\n",
    "from collections import defaultdict\n",
    "import statistics as stats\n",
    "from scipy.spatial import distance\n",
    "from scipy.cluster import hierarchy\n",
    "import itertools\n",
    "from sklearn import linear_model\n",
    "import statsmodels.api as sm\n",
    "from matplotlib.lines import Line2D\n",
    "from matplotlib.collections import PolyCollection\n",
    "from scipy.cluster.hierarchy import dendrogram\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "with open('params_icgcl.yaml', 'rb') as f:\n",
    "    conf = yaml.safe_load(f.read())    # load the config file\n",
    "\n",
    "# plot settings\n",
    "sns.set_style(style='white')\n",
    "\n",
    "plt.rc('text', usetex = True)\n",
    "plt.rc('font', **{'family' : \"sans-serif\"})\n",
    "plt.rc('text.latex', preamble=r\"\\usepackage{amsmath}\"\n",
    "           r\"\\usepackage{amstext}\")\n",
    "plt.rcParams[\"axes.linewidth\"] = 2.50\n",
    "plt.rcParams['xtick.major.size'] = 20\n",
    "plt.rcParams['ytick.major.size'] = 20\n",
    "plt.rcParams['text.usetex'] =  True\n",
    "fsz = 28"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "# Load general settings from the config file\n",
    "save_img = conf['settings']['save_img']\n",
    "dataset_config = conf[\"datasets\"][\"Fleischer\"]\n",
    "\n",
    "# Load dataset-specific parameters\n",
    "input_file_metadata = dataset_config['input_file_metadata']\n",
    "ec_hc_mask_folder = dataset_config['ec_hc_mask_folder']\n",
    "input_file = dataset_config[\"input_file\"]\n",
    "input_folder = dataset_config[\"input_folder\"]\n",
    "output_folder = dataset_config[\"output_folder\"]\n",
    "rep_list = dataset_config[\"rep_list\"]"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data Preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "def data_preprocessing(df,expt_list):\n",
    "    normalized_df = df[['Gene_ID','StartPos','EndPos']]\n",
    "    for expt in expt_list:\n",
    "        normalized_df[expt] = (df[expt])/  df[expt].sum()\n",
    "    normalized_df = normalized_df.reset_index()\n",
    "    return  normalized_df"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plotting Function"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "def setAlpha(ax,a):\n",
    "    for art in ax.get_children():\n",
    "        if isinstance(art, PolyCollection):\n",
    "            art.set_alpha(a)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "def fancy_dendrogram(*args, **kwargs):\n",
    "    max_d = kwargs.pop('max_d', None)\n",
    "    fig, ax = plt.subplots()\n",
    "    fig.set_size_inches(9,6)\n",
    "    fig.set_dpi(150)\n",
    "    if max_d and 'color_threshold' not in kwargs:\n",
    "        kwargs['color_threshold'] = max_d\n",
    "\n",
    "    ddata = dendrogram(*args, **kwargs)\n",
    "\n",
    "    if not kwargs.get('no_plot', False):\n",
    "        plt.xlabel(\"Experiment\",fontsize = fsz+6, color='k')\n",
    "        plt.ylabel(\"Distance\",fontsize = fsz+6, color='k')\n",
    "        plt.tick_params(\"y\",labelsize=fsz-6)\n",
    "        plt.tick_params(\"x\",labelsize=3.5)\n",
    "        plt.tick_params(axis=u'both', which=u'both',length=0)\n",
    "        plt.xlabel('Age (color coded by discretization)')\n",
    "        plt.ylabel('Distance')\n",
    "        if max_d:\n",
    "            plt.axhline(y=max_d, c='k')\n",
    "    return ddata"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# df_sort = = pd.read_csv(os.path.join(output_folder,'fpkm_data_SVD.csv')) # T\n",
    "df_sort = pd.read_csv(os.path.join(output_folder,\"auto_cor_ratio.csv\"))\n",
    "df = df_sort.sort_values('chromosome_id')\n",
    "df = df.set_index('chromosome_id')\n",
    "df = df.T\n",
    "Z = hierarchy.linkage(df, method='average',metric='cityblock',optimal_ordering=True) # Use it to cluster autocorr\n",
    "# Z = hierarchy.linkage(sp.distance.squareform(df), method='average',metric='cityblock',optimal_ordering=True) # Use it to cluster PCA and SVD data"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "D_leaf_colors = {\n",
    "\"101_19YR_F\": \"c\",\n",
    "\"102_19YR_M\": \"c\",\n",
    "\"103_19YR_M\": \"c\",\n",
    "\"104_19YR_M\": \"c\",\n",
    "\"105_20YR_M\": \"c\",\n",
    "\"106_20YR_F\": \"c\",\n",
    "\"14_1YR_M\": \"c\",\n",
    "\"15_12YR_M\": \"c\",\n",
    "\"155_22YR_M\": \"g\",\n",
    "\"156_25YR_F\": \"g\",\n",
    "\"158_24YR_M\": \"g\",\n",
    "\"159_23YR_M\": \"g\",\n",
    "\"162_21YR_M\": \"g\",\n",
    "\"16_24YR_F\": \"g\",\n",
    "\"17_25YR_M\": \"g\",\n",
    "\"18_25YR_F\": \"g\",\n",
    "\"79_1YR_M\": \"c\",\n",
    "\"80_2YR_F\": \"c\",\n",
    "\"81_3YR_M\": \"c\",\n",
    "\"82_3YR_M\": \"c\",\n",
    "\"83_5YR_M\": \"c\",\n",
    "\"84_6YR_M\": \"c\",\n",
    "\"85_7YR_M\": \"c\",\n",
    "\"86_7YR_M\": \"c\",\n",
    "\"87_8YR_M\": \"c\",\n",
    "\"88_8YR_M\": \"c\",\n",
    "\"89_9YR_F\": \"c\",\n",
    "\"90_10YR_M\": \"c\",\n",
    "\"91_10YR_M\": \"c\",\n",
    "\"92_11YR_F\": \"c\",\n",
    "\"93_11YR_F\": \"c\",\n",
    "\"94_12YR_M\": \"c\",\n",
    "\"95_13YR_M\": \"c\",\n",
    "\"97_16YR_F\": \"c\",\n",
    "\"98_17YR_M\": \"c\",\n",
    "\"99_17YR_M\": \"c\",\n",
    "\"107_31YR_F\": \"g\",\n",
    "\"108_31YR_F\": \"g\",\n",
    "\"109_32YR_M\": \"g\",\n",
    "\"110_32YR_F\": \"g\",\n",
    "\"111_33YR_M\": \"g\",\n",
    "\"112_33YR_M\": \"g\",\n",
    "\"114_37YR_F\": \"g\",\n",
    "\"115_37YR_F\": \"g\",\n",
    "\"117_37YR_M\": \"g\",\n",
    "\"118_39YR_M\": \"g\",\n",
    "\"157_29YR_M\": \"g\",\n",
    "\"160_26YR_F\": \"g\",\n",
    "\"19_26YR_M\": \"g\",\n",
    "\"20_26YR_F\": \"g\",\n",
    "\"21_28YR_M\": \"g\",\n",
    "\"22_29YR_M\": \"g\",\n",
    "\"23_29YR_M\": \"g\",\n",
    "\"24_30YR_M\": \"g\",\n",
    "\"25_30YR_F\": \"g\",\n",
    "\"26_30YR_M\": \"g\",\n",
    "\"27_30YR_M\": \"g\",\n",
    "\"28_43YR_F\": \"r\",\n",
    "\"29_44YR_M\": \"r\",\n",
    "\"30_41YR_M\": \"r\",\n",
    "\"31_47YR_M\": \"r\",\n",
    "\"32_50YR_F\": \"r\",\n",
    "\"33_46YR_M\": \"r\",\n",
    "\"34_50YR_M\": \"r\",\n",
    "\"35_42YR_F\": \"r\",\n",
    "\"36_47YR_M\": \"r\",\n",
    "\"37_41YR_F\": \"r\",\n",
    "\"38_43YR_M\": \"r\",\n",
    "\"39_44YR_M\": \"r\",\n",
    "\"40_45YR_M\": \"r\",\n",
    "\"41_46YR_M\": \"r\",\n",
    "\"42_46YR_M\": \"r\",\n",
    "\"43_47YR_M\": \"r\",\n",
    "\"120_51YR_M\": \"r\",\n",
    "\"121_52YR_M\": \"r\",\n",
    "\"122_55YR_M\": \"r\",\n",
    "\"123_57YR_M\": \"r\",\n",
    "\"124_60YR_M\": \"r\",\n",
    "\"125_71YR_F\": \"m\",\n",
    "\"126_75YR_M\": \"m\",\n",
    "\"163_67YR_F\": \"m\",\n",
    "\"164_67YR_F\": \"m\",\n",
    "\"165_68YR_M\": \"m\",\n",
    "\"167_66YR_M\": \"m\",\n",
    "\"168_68YR_M\": \"m\",\n",
    "\"169_70YR_M\": \"m\",\n",
    "\"170_69YR_M\": \"m\",\n",
    "\"44_61YR_M\": \"m\",\n",
    "\"45_62YR_F\": \"m\",\n",
    "\"46_62YR_F\": \"m\",\n",
    "\"47_63YR_M\": \"m\",\n",
    "\"48_64YR_M\": \"m\",\n",
    "\"49_66YR_M\": \"m\",\n",
    "\"50_67YR_M\": \"m\",\n",
    "\"51_67YR_M\": \"m\",\n",
    "\"52_68YR_M\": \"m\",\n",
    "\"53_68YR_M\": \"m\",\n",
    "\"54_69YR_M\": \"m\",\n",
    "\"55_69YR_F\": \"m\",\n",
    "\"127_78YR_M\": \"m\",\n",
    "\"128_80YR_F\": \"m\",\n",
    "\"129_94YR_M\": \"C0\",\n",
    "\"130_89YR_M\": \"C0\",\n",
    "\"131_87YR_M\": \"C0\",\n",
    "\"132_90YR_M\": \"C0\",\n",
    "\"133_89YR_M\": \"C0\",\n",
    "\"134_86YR_M\": \"C0\",\n",
    "\"135_92YR_M\": \"C0\",\n",
    "\"136_87YR_M\": \"C0\",\n",
    "\"166_82YR_F\": \"C0\",\n",
    "\"56_83YR_M\": \"C0\",\n",
    "\"57_83YR_M\": \"C0\",\n",
    "\"58_83YR_M\": \"C0\",\n",
    "\"59_84YR_F\": \"C0\",\n",
    "\"60_84YR_M\": \"C0\",\n",
    "\"61_84YR_F\": \"C0\",\n",
    "\"62_84YR_M\": \"C0\",\n",
    "\"63_84YR_M\": \"C0\",\n",
    "\"64_84YR_M\": \"C0\",\n",
    "\"65_84YR_M\": \"C0\",\n",
    "\"66_86YR_M\": \"C0\",\n",
    "\"67_86YR_M\": \"C0\",\n",
    "\"68_86YR_M\": \"C0\",\n",
    "\"69_87YR_F\": \"C0\",\n",
    "\"70_87YR_M\": \"C0\",\n",
    "\"71_87YR_M\": \"C0\",\n",
    "\"72_87YR_M\": \"C0\",\n",
    "\"73_88YR_M\": \"C0\",\n",
    "\"74_89YR_M\": \"C0\",\n",
    "\"75_90YR_M\": \"C0\",\n",
    "\"76_91YR_M\": \"C0\",\n",
    "\"77_92YR_F\": \"C0\",\n",
    "\"78_96YR_M\": \"C0\",}"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "# notes:\n",
    "# * rows in Z correspond to \"inverted U\" links that connect clusters\n",
    "# * rows are ordered by increasing distance\n",
    "# * if the colors of the connected clusters match, use that color for link\n",
    "link_cols = {}\n",
    "\n",
    "labels = df.index\n",
    "dflt_col = \"#808080\"\n",
    "link_cols = {}\n",
    "for i, i12 in enumerate(Z[:,:2].astype(int)):\n",
    "  c1, c2 = (link_cols[x] if x > len(Z) else D_leaf_colors[labels[x]]\n",
    "    for x in i12)\n",
    "  link_cols[i+1+len(Z)] = c1 if c1 == c2 else dflt_col\n",
    "labels = [sub.replace('_M', '') for sub in labels]\n",
    "labels = [sub.replace('_F', '') for sub in labels]"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "dn = fancy_dendrogram(Z, labels=labels, leaf_rotation=90,color_threshold=.00, link_color_func=lambda x: link_cols[x])\n",
    "\n",
    "ax = plt.gca()\n",
    "ax.grid(False)\n",
    "df_list_treat = {('0','20'): 'c', ('21','40'): 'g', ('41','60'): 'r', ('61','80'): 'm', ('81','100'): 'C0'}\n",
    "labels = ['Age group: '+ list(df_list_treat.items())[0][0][0]+'-'+list(df_list_treat.items())[0][0][1],\n",
    "        'Age group: '+ list(df_list_treat.items())[1][0][0]+'-'+list(df_list_treat.items())[1][0][1],\n",
    "        'Age group: '+ list(df_list_treat.items())[2][0][0]+'-'+list(df_list_treat.items())[2][0][1],\n",
    "        'Age group: '+ list(df_list_treat.items())[3][0][0]+'-'+list(df_list_treat.items())[3][0][1],\n",
    "        'Age group: '+ list(df_list_treat.items())[4][0][0]+'-'+list(df_list_treat.items())[4][0][1]]\n",
    "\n",
    "legend_elements = [Line2D([0], [0],markerfacecolor='None',markeredgecolor='None',color = list(df_list_treat.items())[0][1], label=labels[0],markersize=15),\n",
    "                        Line2D([0], [0],markerfacecolor='None',markeredgecolor='None',color = list(df_list_treat.items())[1][1], label=labels[1],markersize=15),\n",
    "                        Line2D([0], [0],markerfacecolor='None',markeredgecolor='None',color = list(df_list_treat.items())[2][1], label=labels[2],markersize=15),\n",
    "                        Line2D([0], [0],markerfacecolor='None',markeredgecolor='None',color = list(df_list_treat.items())[3][1], label=labels[3],markersize=15),\n",
    "                       Line2D([0], [0],markerfacecolor='None',markeredgecolor='None',color = list(df_list_treat.items())[4][1], label='Age group: $>=$81',markersize=15)]\n",
    "leg = ax.legend(handles=legend_elements,ncol=5, markerscale=3, loc='upper right', borderaxespad=0.05, fontsize = fsz+26, bbox_to_anchor=(0.5, 1.5))\n",
    "leg = ax.legend(handles=legend_elements,markerscale=2, loc='upper right', borderaxespad=0.05, fontsize = fsz-12)\n",
    "for i in leg.legendHandles:\n",
    "    i.set_linewidth(5)\n",
    "x_labels = ax.get_xmajorticklabels()\n",
    "for x in x_labels:\n",
    "    for key in df_list_treat:\n",
    "        key_1 = int(key[0])\n",
    "        key_2 = int(key[1])\n",
    "        x_split =x.get_text().split('_')[1]\n",
    "        s=''\n",
    "        for i in x_split:\n",
    "            if i.isdigit():\n",
    "                s=s+i\n",
    "        x_split = int(s)\n",
    "        if key_1<=x_split<=key_2:\n",
    "           x.set_color(df_list_treat[key])\n",
    "           break\n",
    "# For every axis, set the x and y major locator\n",
    "# ax.xaxis.set_major_locator(plt.MaxNLocator(1))\n",
    "ax.yaxis.tick_right()\n",
    "ax.yaxis.set_label_position(\"right\")\n",
    "plt.tight_layout()\n",
    "# plt.savefig(os.path.join(output_folder,'dendrogram_PCA_sqtranform_withX_1.pdf'),format='pdf', bbox_inches='tight')"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Postprocessing and Analysis"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "# Post processing data analysis beyond this point. Importing the output l* file saved in output_folder directory\n",
    "df = pd.read_csv(os.path.join(output_folder,\"auto_cor_ratio.csv\"))\n",
    "df_chromosome_dist = pd.DataFrame(df.chromosome_id)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "for rep_entry in rep_list:\n",
    "    col_name = ''\n",
    "    if any('19YR' in rep_name for rep_name in rep_entry):\n",
    "        col_name = '0-20'\n",
    "    if any('31YR' in rep_name for rep_name in rep_entry):\n",
    "        col_name = '21-40'\n",
    "    if any('51YR' in rep_name for rep_name in rep_entry):\n",
    "        col_name = '41-60'\n",
    "    if any('71YR' in rep_name for rep_name in rep_entry):\n",
    "        col_name = '61-80'\n",
    "    if any('91YR' in rep_name for rep_name in rep_entry):\n",
    "        col_name = '>=81'\n",
    "    df_chromosome_dist[col_name]=df[rep_entry].mean(axis=1)\n",
    "expt_order = ['0-20', '21-40','41-60','61-80','>=81']"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "## Fig 4a\n",
    "df_chromosome_dist_melted_df = pd.melt(df_chromosome_dist,id_vars='chromosome_id',value_vars=list(df_chromosome_dist.columns))\n",
    "df_chromosome_dist_melted_df=df_chromosome_dist_melted_df.rename(columns={'value':'lstar'})\n",
    "df_chromosome_dist = df_chromosome_dist.set_index('chromosome_id')\n",
    "df_chromosome_dist_viol = pd.DataFrame(index=df_chromosome_dist.index)\n",
    "for col in df_chromosome_dist.columns:\n",
    "    df_chromosome_dist_viol[col] = (df_chromosome_dist[col]).astype(float)\n",
    "df_chromosome_dist_viol = df_chromosome_dist_viol.reset_index()\n",
    "df_chromosome_dist_viol_melted_df = pd.melt(df_chromosome_dist_viol,id_vars='chromosome_id',value_vars=expt_order)\n",
    "df_chromosome_dist_viol_melted_df = df_chromosome_dist_viol_melted_df.rename(columns={'value':'lstar'})\n",
    "fig, ax = plt.subplots()\n",
    "fig.set_size_inches(12, 10)\n",
    "fig.set_dpi(300)\n",
    "# Let's plot the strip plot to check the temp values\n",
    "ax = sns.boxplot(data=df_chromosome_dist_viol_melted_df, x=\"variable\", y=\"lstar\",palette=['c','g','r','m','C0'], dodge = True, linewidth=2, medianprops=dict(linestyle='None'),meanline=True, showmeans=True, meanprops=dict(linestyle='-',linewidth = 2, color='black'), boxprops=dict(alpha=.6))\n",
    "ax.set_xticks(range(len(df_chromosome_dist.columns))) # <--- set the ticks first\n",
    "ax.set_xticklabels(['0-20','21-40','41-60','61-80','$>=$81'])\n",
    "setAlpha(ax,0.6)\n",
    "ax.grid(False)\n",
    "ax.set_ylabel('$\\ell^{*}$', fontsize=fsz+10, color='k')\n",
    "ax.set(xlabel=None)\n",
    "ax.tick_params('both',which='major', length=7,labelsize=fsz)\n",
    "ax.tick_params('both',which='minor', length=7,labelsize=fsz)\n",
    "plt.xticks(rotation=45)\n",
    "plt.show()\n",
    "# plt.tight_layout()\n",
    "# plt.savefig(os.path.join(output_folder,(\"box_plot_ell_age_Fleischer_dbscan\"+\".pdf\")), bbox_inches='tight')"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "# Let's perform Mann Kendall test for the above data sets to identify the trend of the curves\n",
    "y_data = []\n",
    "for value in df_chromosome_dist_viol_melted_df['variable'].unique():\n",
    "    subset = df_chromosome_dist_viol_melted_df[df_chromosome_dist_viol_melted_df['variable'] == value]\n",
    "    y_data.append(subset['lstar'].mean()) # Using it mean first and then we can test for median    \n",
    "result = mk.original_test(y_data)\n",
    "print(result)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "## Fig 4b\n",
    "# Now we can melt the data\n",
    "df = df.set_index('chromosome_id')\n",
    "df_chromosome_dist_viol = pd.DataFrame(index=df.index)\n",
    "for col in df.columns:\n",
    "    df_chromosome_dist_viol[col] = (df[col]).astype(float) \n",
    "df_chromosome_dist_viol = df_chromosome_dist_viol.reset_index()\n",
    "df_chromosome_dist_viol_melted_df = pd.melt(df_chromosome_dist_viol, id_vars='chromosome_id')\n",
    "df_chromosome_dist_viol_melted_df = df_chromosome_dist_viol_melted_df.rename(columns={'value': 'lstar'})\n",
    "# Let's now replace lists with their respective age\n",
    "df_chromosome_dist_viol_melted_df['Age'] = df_chromosome_dist_viol_melted_df['variable'].str.extract(r'(\\d+)(?=YR)')"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "# Using the linear regression model to measure the linear relationship between age and ell*\n",
    "features = ['Age']\n",
    "target = 'lstar'\n",
    "df_chromosome_dist_viol_melted_df['chromosome_id'] = df_chromosome_dist_viol_melted_df['chromosome_id'].replace('X', 23)\n",
    "X = df_chromosome_dist_viol_melted_df[features].values.reshape(-1, len(features))\n",
    "y = df_chromosome_dist_viol_melted_df[target].values\n",
    "ols = linear_model.LinearRegression()\n",
    "model = ols.fit(X, y)\n",
    "\n",
    "X = [int(item[0]) for item in X]\n",
    "X2 = sm.add_constant(X)\n",
    "model1 = sm.OLS(y, X2).fit()\n",
    "print(model1.summary())"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "y_data = {}\n",
    "y_data_std = {}\n",
    "for value in df_chromosome_dist_viol_melted_df['Age'].unique():\n",
    "    subset = df_chromosome_dist_viol_melted_df[df_chromosome_dist_viol_melted_df['Age'] == value]\n",
    "    y_data[value] = subset['lstar'].mean()\n",
    "    y_data_std[value] = subset['lstar'].std()\n",
    "\n",
    "sorted_data = {k: y_data[k] for k in sorted(y_data, key=int)}\n",
    "sorted_std = {k: y_data_std[k] for k in sorted(y_data_std, key=int)}\n",
    "sorted_keys = list(sorted_data.keys())\n",
    "values = list(sorted_data.values())\n",
    "values_std = list(sorted_std.values())\n",
    "result = values\n",
    "print(mk.original_test(result))"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "x_pred = np.linspace(0, len(sorted_keys)+1, 20)     \n",
    "x_pred_1 = np.linspace(0, 98, 20)     \n",
    "x_pred = x_pred.reshape(-1, len(features))  \n",
    "x_pred_1 = x_pred_1.reshape(-1, len(features))  \n",
    "y_pred = model.predict(x_pred)\n",
    "y_pred_1 = model.predict(x_pred_1)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "# Function to find the color based on the group\n",
    "def get_color(label, group_colors):\n",
    "    for group, color in group_colors.items():\n",
    "        if int(group[0]) <= label <= int(group[1]):\n",
    "            return color\n",
    "    return 'black'"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "# Plot th sorted keys against their corresponding values\n",
    "fig, axs = plt.subplots()\n",
    "fig.set_size_inches(14, 7)\n",
    "fig.set_dpi(300)\n",
    "\n",
    "axs.plot(sorted_keys,result,'ok',linewidth = 3)\n",
    "axs.errorbar(sorted_keys, values, yerr=values_std, fmt='o', markersize=8,markeredgecolor='k', capsize=10, color=\"k\", elinewidth=2, markeredgewidth=2)\n",
    "axs.plot(x_pred, y_pred, color='k',linewidth = 2, label='Regression model')\n",
    "plt.xlabel('Age of the donor', fontsize=fsz, color='k')\n",
    "plt.ylabel(r'$\\ell^{*}$', size=fsz+4)\n",
    "plt.xticks(rotation=45)  # Rotate x-axis labels for better readability\n",
    "axs.tick_params('both',which='major', length=7,labelsize=fsz)\n",
    "axs.tick_params('both',which='minor', length=7,labelsize=fsz)\n",
    "axs.tick_params(axis='x', rotation=90, labelsize=fsz-10)\n",
    "\n",
    "# Example groupings and corresponding colors\n",
    "group_colors = {\n",
    "    ('0', '20'): 'c',\n",
    "    ('21', '40'): 'g',\n",
    "    ('41', '60'): 'r',\n",
    "    ('61', '80'): 'm',\n",
    "    ('81', '100'): \"C0\"\n",
    "}\n",
    "sorted_keys = list(map(int, sorted_keys))\n",
    "# Set the color for each tick label based on the group\n",
    "for key, i in enumerate(sorted_keys):\n",
    "    color = get_color(i, group_colors)\n",
    "    axs.get_xticklabels()[key].set_color(color)\n",
    "plt.show()\n",
    "plt.tight_layout()\n",
    "# plt.savefig(os.path.join(output_folder,'l_star_vs_age_Fleischer_ec_gamma3.pdf'), format='pdf', bbox_inches='tight')"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# # Plot\n",
    "# fig, axs = plt.subplots(figsize=(14, 7), dpi=300)\n",
    "# \n",
    "# # Plot only points that have values (ignore zeros)\n",
    "# axs.plot(sorted_keys,result, 'ok', linewidth=3)\n",
    "# axs.errorbar(sorted_keys, values, yerr=values_std, fmt='o', \n",
    "#              markersize=8, markeredgecolor='k', capsize=10, color=\"k\", \n",
    "#              elinewidth=2, markeredgewidth=2)\n",
    "# \n",
    "# # X-axis labels (show only every 10th tick)\n",
    "# axs.set_xticks(np.arange(0, 101, 10))\n",
    "# axs.set_xticklabels(np.arange(0, 101, 10), fontsize=12)\n",
    "# \n",
    "# axs.plot(x_pred, y_pred, color='k',linewidth = 2, label='Regression model')\n",
    "# plt.xlabel('Age of the donor', fontsize=fsz, color='k')\n",
    "# plt.ylabel(r'$\\ell^{*}$', size=fsz+4)\n",
    "# plt.xticks(rotation=45)  # Rotate x-axis labels for better readability\n",
    "# axs.tick_params('both',which='major', length=7,labelsize=fsz)\n",
    "# axs.tick_params('both',which='minor', length=7,labelsize=fsz)\n",
    "# axs.tick_params(axis='x', rotation=45, labelsize=fsz-6)\n",
    "# \n",
    "# # Example groupings and corresponding colors\n",
    "# plt.show()\n",
    "# plt.tight_layout()"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Define group colors based on age ranges\n",
    "group_colors = {\n",
    "    ('0', '20'): 'c',\n",
    "    ('21', '40'): 'g',\n",
    "    ('41', '60'): 'r',\n",
    "    ('61', '80'): 'm',\n",
    "    ('81', '100'): \"C0\"\n",
    "}\n",
    "\n",
    "# Function to get color for an age group\n",
    "def get_color(age, group_colors):\n",
    "    for (low, high), color in group_colors.items():\n",
    "        if int(low) <= age <= int(high):\n",
    "            return color\n",
    "    return 'k'  # Default color if none match\n",
    "\n",
    "# Plot\n",
    "fig, axs = plt.subplots()\n",
    "fig.set_size_inches(14, 7)\n",
    "fig.set_dpi(300)\n",
    "# Plot the points\n",
    "axs.plot(sorted_keys, values, 'ok', linewidth=3)\n",
    "\n",
    "# Add error bars with colors based on group\n",
    "for i, key in enumerate(sorted_keys):\n",
    "    color = get_color(key, group_colors)\n",
    "    axs.errorbar(key, values[i], yerr=values_std[i], fmt='o', markersize=8, \n",
    "                 markeredgecolor='k', capsize=10, color=color, \n",
    "                 elinewidth=2, markeredgewidth=2)\n",
    "\n",
    "# X-axis labels (show only every 10th tick)\n",
    "axs.set_xticks(np.arange(0, 101, 10))\n",
    "axs.set_xticklabels(np.arange(0, 101, 10), fontsize=12)\n",
    "\n",
    "axs.plot(x_pred_1, y_pred_1, color='k',linewidth = 2, label='Regression model')\n",
    "plt.xlabel('Age of the donor', fontsize=fsz, color='k')\n",
    "plt.ylabel(r'$\\ell^{*}$', size=fsz+4)\n",
    "plt.xticks(rotation=45)  # Rotate x-axis labels for better readability\n",
    "axs.tick_params('both',which='major', length=7,labelsize=fsz)\n",
    "axs.tick_params('both',which='minor', length=7,labelsize=fsz)\n",
    "axs.tick_params(axis='x', rotation=45, labelsize=fsz-6)\n",
    "\n",
    "# Example groupings and corresponding colors\n",
    "# plt.show()\n",
    "plt.tight_layout()\n",
    "plt.savefig(os.path.join(output_folder,'l_star_vs_age_Fleischer_ec_gamma3.pdf'), format='pdf', bbox_inches='tight')"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "# Reading the files containing HAT and LAT genes. The genes labeled 1 are HATs and genes labeled 0 are LATs\n",
    "df_ec_hc = pd.DataFrame()\n",
    "counter = 0\n",
    "for chromosome_dir in os.listdir(ec_hc_mask_folder):\n",
    "    f = os.path.join(ec_hc_mask_folder, chromosome_dir)\n",
    "    if os.path.isdir(f):\n",
    "        for filename in os.listdir(f):\n",
    "            if filename.endswith(\"_ec_hc_mask_xy.csv\"):  # Check for files that match the pattern\n",
    "                file_path = os.path.join(f, filename)  # Get the full path of the file\n",
    "                df_ec_hc_temp = pd.read_csv(file_path)\n",
    "                df_ec_hc_temp.drop(['StartPos','EndPos'],axis=1,inplace=True)\n",
    "                df_ec_hc_temp = df_ec_hc_temp.set_index('Gene_ID')\n",
    "                if counter == 0:\n",
    "                    df_ec_hc = df_ec_hc_temp\n",
    "                else:\n",
    "                    df_ec_hc = df_ec_hc.add(df_ec_hc_temp, fill_value=0)\n",
    "                counter = counter+1\n",
    "df_ec_hc['sum'] = df_ec_hc.sum(axis=1)\n",
    "df_ec_hc = df_ec_hc[df_ec_hc['sum']>0]\n",
    "index_ec_genes = list(df_ec_hc.index)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "## Fig 3c\n",
    "expt_list = [item for sublist in rep_list for item in sublist]\n",
    "df.sort_values(by=['chromosome_id'],inplace=True)\n",
    "df_vec = df.reset_index(drop=True)\n",
    "df_list = df_vec.loc[:, df.columns != 'chromosome_id'].to_numpy()\n",
    "dist_df = pd.DataFrame(columns= ['reference'] + expt_list)\n",
    "ref_list = rep_list[0]\n",
    "dist_df['reference'] = ref_list\n",
    "for i in range(0,len(rep_list)):\n",
    "    for rep1, rep2 in itertools.product(ref_list,rep_list[i]):\n",
    "        dist = distance.cityblock(df_vec[rep1],df_vec[rep2])\n",
    "        dist_df.loc[dist_df['reference'] == rep1,rep2] = dist\n",
    "label_list = ['Grp_1_20','Grp_21_40','Grp_41_60','Grp_61_80','Grp_81_100']\n",
    "# Setting up the mean and standard deviation plot\n",
    "dist_list = []\n",
    "dist_list_std = []\n",
    "dist_list_df = pd.DataFrame(columns= label_list)\n",
    "dist_list_df_std = pd.DataFrame(columns= label_list)\n",
    "for rep in rep_list:\n",
    "    dist_list.append(np.average(dist_df[rep].mean().to_numpy()))\n",
    "    dist_list_std.append(np.std(dist_df[rep].to_numpy()))\n",
    "\n",
    "dist_list_df = dist_list_df.append(pd.DataFrame([dist_list], columns= label_list), ignore_index=True)\n",
    "dist_list_df_std = dist_list_df_std.append(pd.DataFrame([dist_list_std], columns= label_list), ignore_index=True)\n",
    "# Setting up the plotting tool\n",
    "df_sim_auto_corr = pd.concat([dist_list_df,dist_list_df_std])\n",
    "df_sim_auto_corr = df_sim_auto_corr.T.reset_index()\n",
    "df_sim_auto_corr.columns = ['Expt','mean','std']\n",
    "x_col = ['Grp_1_20','Grp_21_40','Grp_41_60','Grp_61_80','Grp_81_100']\n",
    "df_sim_auto_corr['Expt'] = pd.Categorical(df_sim_auto_corr.Expt, categories=x_col, ordered=True)\n",
    "df_sim_auto_corr =df_sim_auto_corr.sort_values('Expt')\n",
    "## Let's save these values as well that is required to do the plotting\n",
    "# df_sim_auto_corr.to_csv(os.path.join(output_folder,\"pair_wise_similarity_autocorr.csv\"),index=False)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "# Let's test another algorithm\n",
    "df = pd.read_csv(input_file)\n",
    "result = defaultdict(list)\n",
    "dist_list = []\n",
    "dist_list_std = []\n",
    "\n",
    "for i in range(0,len(rep_list)):\n",
    "    result = []\n",
    "    for rep1, rep2 in itertools.product(ref_list,rep_list[i]):\n",
    "        dist = 0\n",
    "        counter = 0\n",
    "        for chromosome,data in df.groupby('Chromosome'):\n",
    "            if chromosome.isalnum() and chromosome != 'MT':\n",
    "                counter = counter + 1\n",
    "                normalized_df = data_preprocessing(data,[rep1,rep2])\n",
    "                normalized_df = normalized_df[normalized_df['Gene_ID'].isin(index_ec_genes)].reset_index(drop=True) # Filter out EC genes\n",
    "                # Let's calculate the similarity distance between the two distributions\n",
    "                dist = dist + distance.cityblock(normalized_df[rep1],normalized_df[rep2])\n",
    "        avg_dist = dist/counter\n",
    "        result.append(avg_dist)\n",
    "    dist_list.append(stats.mean(result))\n",
    "    dist_list_std.append(stats.stdev(result))\n",
    "    "
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "# Setting up the saving the data\n",
    "df_sim_exp = pd.DataFrame([dist_list,dist_list_std],columns = label_list)\n",
    "df_sim_exp = df_sim_exp.T.reset_index()\n",
    "df_sim_exp.columns = ['Expt','mean','std']\n",
    "x_col = ['Grp_1_20','Grp_21_40','Grp_41_60','Grp_61_80','Grp_81_100']\n",
    "df_sim_exp['Expt'] = pd.Categorical(df_sim_exp.Expt, categories=x_col, ordered=True)\n",
    "df_df_sim_exp = df_sim_exp.sort_values('Expt')\n",
    "## Let's save these values as well that is required to do the plotting\n",
    "# df_df_sim_exp.to_csv(os.path.join(output_folder,\"pair_wise_similarity_expression.csv\"),index=False)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "y_exp = df_sim_exp['mean'].to_list()\n",
    "yerr_exp = df_sim_exp['std'].to_list()\n",
    "y_autocorr = df_sim_auto_corr['mean'].to_list()\n",
    "yerr_autocorr = df_sim_auto_corr['std'].to_list()\n",
    "x_label = df_sim_auto_corr['Expt'].to_list()\n",
    "\n",
    "x1 = [1,2,3,4,5]\n",
    "# Dictionary mapping original labels to new labels\n",
    "label_mapping = {\n",
    "    'Grp_1_20': '0-20',\n",
    "    'Grp_21_40': '21-40',\n",
    "    'Grp_41_60': '41-60',\n",
    "    'Grp_61_80': '61-80',\n",
    "    'Grp_81_100': '81-100'\n",
    "}\n",
    "# Updating the list using the mapping\n",
    "x_label = [label_mapping[group] for group in x_label]\n",
    "# Setting up the plotting tool\n",
    "fsz = 28\n",
    "fig, ax = plt.subplots()\n",
    "fig.set_size_inches(12, 8)\n",
    "fig.set_dpi(300)\n",
    "plt.plot(x1,y_exp,'-b',linewidth = 3)\n",
    "plt.errorbar(x1, y_exp, yerr=yerr_exp, fmt='o', markersize=8, capsize=10,markeredgecolor='b', color=\"b\", elinewidth=2, markeredgewidth=2)\n",
    "plt.xticks(rotation=45)\n",
    "ax1 = ax.twinx()\n",
    "ax1.plot(x1,y_autocorr,'-r',linewidth = 3)\n",
    "ax1.errorbar(x1, y_autocorr, yerr=yerr_autocorr, fmt='o', markersize=8,markeredgecolor='r', capsize=10, color=\"r\", elinewidth=2, markeredgewidth=2)\n",
    "\n",
    "ax.set_ylabel(r'$\\mathcal{D}(\\alpha_t)$', fontsize=fsz+10, color='k')\n",
    "ax1.set_ylabel(r'$\\mathcal{D}(\\ell^{*})$', fontsize=fsz+10, color='k')\n",
    "ax.set_xlim(left=0.9,right=5.1)\n",
    "ax1.set_ylim(bottom=0,top=70)\n",
    "ax.set_ylim(bottom=0,top=0.6)\n",
    "\n",
    "ax1.locator_params(axis = 'y', nbins=5)\n",
    "ax.locator_params(axis = 'y', nbins=5)\n",
    "\n",
    "ax.set_xticks(x1)\n",
    "ax.set_xticklabels(x_label)\n",
    "\n",
    "# Setting the color for the axis\n",
    "ax.yaxis.label.set_color('b')\n",
    "ax1.spines[\"left\"].set_edgecolor('b')\n",
    "ax.tick_params(axis='y', colors='b')\n",
    "ax1.yaxis.label.set_color('r')\n",
    "ax1.spines[\"right\"].set_edgecolor('r')\n",
    "ax1.tick_params(axis='y', colors='r')\n",
    "ax.tick_params('both',which='major', length=7,labelsize=fsz)\n",
    "ax.tick_params('both',which='minor', length=7,labelsize=fsz)\n",
    "ax1.tick_params('both',which='major', length=7,labelsize=fsz)\n",
    "ax1.tick_params('both',which='minor', length=7,labelsize=fsz)\n",
    "plt.tight_layout()\n",
    "# plt.savefig(os.path.join(output_folder,(\"pair_wise_similarity_Fleischer\"+\".pdf\")), bbox_inches='tight')"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "## Fig S11\n",
    "df =  pd.read_csv(os.path.join(output_folder,\"auto_cor_ratio.csv\"))\n",
    "df_md = pd.read_csv(input_file_metadata)\n",
    "## Setting up the sorting code for chromosome Id\n",
    "sorter = [str(x) for x in range(1,23)]\n",
    "sorter.append('X')\n",
    "sorter.append('Y')\n",
    "sorterIndex = dict(zip(sorter, range(len(sorter))))\n",
    "df.sort_values(by=['chromosome_id'], key=lambda x: x.map(sorterIndex), inplace = True)\n",
    "## Let's combine the information of sample using metadata and plot it\n",
    "chromosomes=df['chromosome_id']\n",
    "df1 = df.iloc[:,1:]\n",
    "df_final = pd.DataFrame()\n",
    "for samplename in df1.columns.tolist():\n",
    "    tmpdf = pd.DataFrame(data={'lstar' : df1[samplename],'sampleNames':samplename,'chromosome':chromosomes})\n",
    "    df_final = pd.concat([df_final, tmpdf], ignore_index=True)\n",
    "df_final = df_md.merge(df_final)\n",
    "# Let's write the analysis to plot a bar plot for different chromosomes one on top of each other\n",
    "df_top = df_final[df_final['chromosome'].isin(sorter[0:12])]\n",
    "df_down = df_final[df_final['chromosome'].isin(sorter[12:])]\n",
    "fig, axes = plt.subplots(2, 1, figsize=(22, 8))\n",
    "custom_palette = sns.color_palette([\"c\", \"g\", \"r\", \"m\", \"C0\"])\n",
    "sns.barplot(ax = axes[0],data = df_top,x='chromosome',y='lstar',hue='age_group',\n",
    "            palette=custom_palette,\n",
    "            capsize = 0.05,             \n",
    "            saturation = 8,    \n",
    "            alpha = 0.8,\n",
    "            errcolor = 'gray', errwidth = 2,  \n",
    "            ci = 'sd'   \n",
    "            )\n",
    "axes[0].set_ylim(0, 22)\n",
    "axes[0].set_ylabel('$\\ell^{*}$', fontsize=fsz+10, color='k')\n",
    "axes[0].set_xlabel('chromosome', fontsize=fsz+10, color='k')\n",
    "axes[0].tick_params('both',which='major', length=7,labelsize=fsz)\n",
    "axes[0].tick_params('both',which='minor', length=7,labelsize=fsz)\n",
    "axes[0].tick_params(axis=u'both', which=u'both',length=0)\n",
    "axes[0].legend(loc='center left', bbox_to_anchor=(1, 0.5), fontsize = fsz-10)\n",
    "sns.barplot(ax = axes[1],data = df_down,x='chromosome',y='lstar',hue='age_group',\n",
    "            palette = custom_palette,\n",
    "            capsize = 0.05,             \n",
    "            saturation = 8,\n",
    "            alpha = 0.8,\n",
    "            errcolor = 'gray', errwidth = 2,  \n",
    "            ci = 'sd'   \n",
    "            )\n",
    "axes[1].set_ylim(0, 22)\n",
    "axes[1].legend(loc='center left', bbox_to_anchor=(1, 0.5), fontsize = fsz-10)\n",
    "axes[1].set_ylabel('$\\ell^{*}$', fontsize=fsz+10, color='k')\n",
    "axes[1].set_xlabel('chromosome', fontsize=fsz+10, color='k')\n",
    "axes[1].tick_params('both',which='major', length=7,labelsize=fsz)\n",
    "axes[1].tick_params('both',which='minor', length=7,labelsize=fsz)\n",
    "axes[1].tick_params(axis=u'both', which=u'both',length=0)\n",
    "# plt.savefig(os.path.join(output_folder,'Fig1_chromosome_lstar_Fleischer.pdf'), format='pdf', bbox_inches='tight')\n",
    "plt.show()"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "## Fig S2 c and d\n",
    "def fit_distribution_2d_comp(normalized_df, expt_name):\n",
    "    # Create subplots\n",
    "    fig, axs = plt.subplots(1, len(expt_name), figsize=(14, 6), sharex=True, sharey=True)\n",
    "    i = 0\n",
    "    title_label = [\"51YR\",\"94YR\"]\n",
    "    for expt in expt_name:\n",
    "        print(expt)\n",
    "        m12_gid = mask_6.loc[mask_6[expt] == 1]['Gene_ID'].tolist()\n",
    "        temp = normalized_df.loc[normalized_df.Gene_ID.isin(m12_gid)]\n",
    "        dpgmm = mixture.BayesianGaussianMixture(n_components=10, covariance_type='full', random_state=10)\n",
    "        temp['cluster_' + expt] = dpgmm.fit_predict(temp[['index', expt]])\n",
    "\n",
    "        gmm = mixture.GaussianMixture(n_components=len(set(temp['cluster_' + expt])), covariance_type='full').fit(temp[['index', expt]])\n",
    "        x_min = normalized_df['index'].min()\n",
    "        x_max = normalized_df['index'].max()\n",
    "\n",
    "        x_range = np.linspace(x_min, x_max, len(normalized_df))\n",
    "        y_range = np.linspace(normalized_df[expt].min(), normalized_df[expt].max(), len(normalized_df))\n",
    "        X, Y = np.meshgrid(x_range, y_range)\n",
    "\n",
    "        pdf = np.exp(gmm._estimate_log_prob(np.column_stack([X.ravel(), Y.ravel()])))\n",
    "        overall_pdf = np.sum(pdf, axis=1)\n",
    "\n",
    "        # Determine the position in the 2x2 grid\n",
    "        ax = axs[i]\n",
    "\n",
    "        # Plot the distributions\n",
    "        ax.plot(x_range, overall_pdf[:len(x_range)], linestyle='--', linewidth=2)\n",
    "        ax.plot(x_range, pdf[:len(x_range)])\n",
    "        ax.set_title(title_label[i], fontsize=fsz - 8)\n",
    "        ax.tick_params('both',which='major', length=7,labelsize=fsz-8)\n",
    "        ax.tick_params('both',which='minor', length=7,labelsize=fsz-8)\n",
    "        i = i + 1\n",
    "\n",
    "    fig.text(0.5, 0.04, 'Genomic coordinates sorted by their position', ha='center', fontsize=fsz-10, color='k')\n",
    "    fig.text(0.04, 0.5, 'Probability distribution of gene expression', va='center', rotation='vertical', fontsize=fsz-10, color='k')\n",
    "\n",
    "    plt.tight_layout(rect=[0.05, 0.05, 1, 0.95])\n",
    "    # plt.savefig(os.path.join(output_folder,(\"P_dist_genomic_coord_chrom6_gamma3\"+\".pdf\")), bbox_inches='tight')\n",
    "    plt.show()"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "def create_ranges(diff):\n",
    "    ranges = []\n",
    "    start = 1\n",
    "    end = diff\n",
    "    max = 100\n",
    "    while end <= max: # assuming the upper limit is 100\n",
    "        ranges.append((start, end))\n",
    "        start = end + 1 # Change it to 1\n",
    "        end = start + diff - 1 # Uncomment this before 1\n",
    "    if start <= max:\n",
    "        ranges.append((start, max))\n",
    "    return ranges"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "def create_rep_list(expt_list,dx):\n",
    "    ranges = create_ranges(dx)\n",
    "    sublists = [[] for _ in range(math.ceil(100/dx))]  # create empty sublists for each age range\n",
    "    for item in expt_list:\n",
    "        age_range = int(item.split('_')[1].replace('YR', ''))\n",
    "        for i in range(len(ranges)):\n",
    "            if ranges[i][0] <= age_range <= ranges[i][1]:\n",
    "                sublists[i].append(item)\n",
    "    return sublists"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "df = pd.read_csv(input_file)\n",
    "mask_6 = pd.read_csv(os.path.join(ec_hc_mask_folder, \"chromosome_6\", 'chromosome_6_ec_hc_mask_xy.csv'))\n",
    "dx = 20\n",
    "rep_list = create_rep_list(expt_list, dx)\n",
    "\n",
    "for chromosome, data in df.groupby('Chromosome'):\n",
    "    expt_list = [item for sublist in rep_list for item in sublist]\n",
    "    if (chromosome.isdigit() and int(chromosome) == 6):\n",
    "        data.sort_values(by=['StartPos', 'EndPos'], inplace=True)\n",
    "        chromosome_output_folder = os.path.join(output_folder, \"chromosome_\" + str(chromosome))\n",
    "        Path(chromosome_output_folder).mkdir(parents=True, exist_ok=True)\n",
    "        normalized_df = data_preprocessing(data, expt_list)\n",
    "        normalized_df = normalized_df.drop(['index'], axis=1)\n",
    "        normalized_df = normalized_df.reset_index()\n",
    "        break"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "# expt_name_comp = [\"99_17YR_M\",\"120_51YR_M\",\"126_75YR_M\",\"129_94YR_M\"]\n",
    "expt_name_comp = [\"120_51YR_M\",\"129_94YR_M\"]\n",
    "fit_distribution_2d_comp(normalized_df,expt_name_comp)"
   ],
   "outputs": [],
   "execution_count": null
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "research-venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
